src/net/tls.cc

```diff
@@ -1369,6 +1369,18 @@ class session : public enable_lw_shared_from_this<session> {
                return put(std::move(p));
             });
         }
+
+        // We want to make sure that we call gnutls_record_send with as large
+        // packets as possible. This is because each call to gnutls_record_send
+        // translates to a sendmsg syscall. Further it results in larger TLS
+        // records which makes encryption/decryption faster. Hence to avoid
+        // cases where we would do an extra syscall for something like a 100
+        // bytes header we linearize the packet if it's below the max TLS record
+        // size.
+        if (p.nr_frags() > 1 && p.len() <= 16384) {
```

Comment:
Nit, 16k really should be a constant somewhere, but the comment more or less makes up for it. :-)

Reply:
16k isn't small, but maybe for TLS it's reasonable.

Is it possible to use a vectored version of gnutls_record_send? I couldn't find one.

Another option is to change vec_push() to queue the data locally, then send it all at once. But that keeps the extra TLS framing, so it's not a good option.

Reply:
I see gnutls_record_cork() and gnutls_record_uncork(). But I suspect it will generate extra frames, or perhaps it will linearize internally.

Reply:
Looked at the code, cork linearizes. So as far as I can tell this code is the best we can do.

Reply:
re the constant, since it's used once adding a name doesn't really help.

Reply:
(but using one from gnutls will)

Reply:
> Is it possible to use a vectored version of gnutls_record_send? I couldn't find one.

Yeah don't think there is.

> Another option is to change vec_push() to queue the data locally, then send it all at once. But that keeps the extra TLS framing, so it's not a good option.

Yes that is another further improvement but is a bit of larger change. This is being complicated by how do_put and the push function currently interact via `wait_for_output`. 

> I see gnutls_record_cork() and gnutls_record_uncork(). But I suspect it will generate extra frames, or perhaps it will linearize internally.

Yeah corking does linearize internally so it's fairly similar to what this patch does. However I think the solution in this patch is easier as it again avoids any complications introduced by corking in relation to wait_for_output. 

> (but using one from gnutls will)

Done now. I changed it so that we use the actual session such that in the unlikely case of a smaller value being negotiated that is being respected now.